### PR TITLE
Fix and add random tests for `ChebyshevPolynomial`

### PR DESCRIPTION
**Pull request by charlesyuan314**
- Switch from `GlobalPhase` back to `ReflectionUsingPrepare` for the correct reflection operator
- Add tests generated from random Hermitian matrices of norm <= 1.

### CODE REVIEW

**Review by fdmalone — APPROVED** [review]

File: qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py [hunk 1]
```diff
@@ -144,3 +151,80 @@ def test_scaled_chebyshev_odd_tensors():
...
+        raise NotImplementedError
+
+    @property
+    def selection_registers(self) -> Tuple[Register, ...]:
+        raise NotImplementedError
+
+    def build_composite_bloq(
+        self, bb: BloqBuilder, system: Soquet, ancilla: Soquet, resource: Soquet
+    ) -> Dict[str, SoquetT]:
+        bits = bb.join(np.concatenate([bb.split(system), bb.split(ancilla), bb.split(resource)]))
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py:193]: are the splits here necessary?

File: qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py [hunk 2]
```diff
@@ -144,3 +151,80 @@ def test_scaled_chebyshev_odd_tensors():
...
+def test_chebyshev_matrix():
+    a = (XGate().tensor_contract() + 3.0 * Hadamard().tensor_contract()) / 4.0
+    from_gate = t4(a)
+    bloq = ChebyshevPolynomial(TestBlockEncoding(block_encoding_matrix(a)), order=4)
+    from_tensors = gate_test(bloq)
+    np.testing.assert_allclose(from_gate, from_tensors, atol=2e-15)
+
+
+def random_hermitian_matrix(random_state):
+    a = random_orthogonal(2, random_state=random_state)
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py:218]: is this a 2x2 matrix? if so how is it working with a 3-qubit matrix gate?

Comment (anurudhp) [reply qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py:218]: nvm I missed the wrapper

Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py:218]: `block_encoding_matrix` will first embed it into a 4x4 matrix and then tensor that with another identity. The result is 8x8.

File: qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py (hunk 1 above)
Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py]: They were not. I have removed them.

File: qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py [hunk 3]
```diff
@@ -144,3 +151,80 @@ def test_scaled_chebyshev_odd_tensors():
...
+
+def test_chebyshev_matrix():
+    a = (XGate().tensor_contract() + 3.0 * Hadamard().tensor_contract()) / 4.0
+    from_gate = t4(a)
+    bloq = ChebyshevPolynomial(TestBlockEncoding(block_encoding_matrix(a)), order=4)
+    from_tensors = gate_test(bloq)
+    np.testing.assert_allclose(from_gate, from_tensors, atol=2e-15)
+
+
+def random_hermitian_matrix(random_state):
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py:217]: this seems useful in general. Maybe add a docstring, and make dimension a parameter?

Could also move it to `qualtran/linalg/matrix.py` or sth

File: qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py [hunk 4]
```diff
@@ -144,3 +151,80 @@ def test_scaled_chebyshev_odd_tensors():
...
+    def build_composite_bloq(
+        self, bb: BloqBuilder, system: Soquet, ancilla: Soquet, resource: Soquet
+    ) -> Dict[str, SoquetT]:
+        bits = bb.join(np.concatenate([bb.split(system), bb.split(ancilla), bb.split(resource)]))
+        bits = bb.add(MatrixGate(3, self.matrix, atol=3e-8), q=bits)
+        system, ancilla, resource = bb.split(bits)
+        return {"system": system, "ancilla": ancilla, "resource": resource}
+
+
+def block_encoding_matrix(a):
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/chebyshev_polynomial_test.py:199]: perhaps this should be a factory constructor for `TestBlockEncoding` as the behaviour seems tightly coupled (for now)

File: qualtran/bloqs/block_encoding/chebyshev_polynomial.py [hunk 5]
```diff
@@ -129,23 +128,23 @@ def signal_state(self) -> PrepareOracle:
...
+            bitsizes=(self.ancilla_bitsize,), global_phase=-1
         )
 
     def build_composite_bloq(self, bb: BloqBuilder, **soqs: SoquetT) -> Dict[str, SoquetT]:
         if is_symbolic(self.ancilla_bitsize):
             raise DecomposeTypeError(f"Cannot decompose symbolic {self=}")
         for _ in range(self.order // 2):
             if self.ancilla_bitsize > 0:
-                soqs["ancilla"] = bb.add(self.reflection_bloq, ctrl=soqs["ancilla"])
+                soqs["ancilla"] = bb.add(self.reflection_bloq, reg0_=soqs["ancilla"])
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/chebyshev_polynomial.py:140]: This seems fragile, would it break if we used a different prepare bloq? Or rather, is the prepare bloq configurable by the user or always picked to be the same one?

Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/chebyshev_polynomial.py:140]: What do you mean by different prepare bloq? the `reg0_` name is dictated by the `ReflectionUsingPrepare`. The reflection bloq in `ChebyshevPolynomial` is not user-configurable.

Comment (anurudhp) [reply qualtran/bloqs/block_encoding/chebyshev_polynomial.py:140]: Ah nvm, I thought this was reflecting around the signal state.

**Review by anurudhp — APPROVED** [review]
lgtm

**Pull request merged**